src/main/groovy/net/vivin/gradle/versioning/VersionUtils.groovy

```diff
@@ -140,7 +144,37 @@ class VersionUtils {
             refresh()
         }
 
-        return versions?.find()
+        if(!versionByTag) {
+            return null
+        } else {
+            def revWalk = new RevWalk(repository)
+            def candidateTags = []
+            def toInvestigateForTags = []
+
+            try {
+                toInvestigateForTags.add(revWalk.parseCommit(repository.resolve(Constants.HEAD)))
+
+                for(RevCommit investigatee = toInvestigateForTags.pop();
```

Comment:
Well, sometimes, especially when the for-parts get too long I tend to format it in three lines to make it better readable.

Actually it is depth-first while I had it breadth-first in the Java version, but whether depth-first or breadth-first shouldn't make a difference in this case, as I do not stop until all is traversed and i sort the result, so the depth-first and breadth-first should produce the same result in this case.

But why should a `while` be more idiomatic? A variable is initialized, a condition is checked before each invocation and after each invocation the variable is updated. This is exactly what a for-loop is meant for. I'd only consider using a while-loop in such a situation if I need the value of the variable after the loop. But even then I would still use a for-loop if at the end of each iteration some task (variable assignment) should be done. If you make it a while loop, you have to put the variable initialization before the loop and thus broaden its scope unnecessarily and it could happen that someone writes code after the between-iteration-task (variable assignment) as you have to put it inside the loop.